### PR TITLE
Add a sample for service-orchestration-with-parallel-processing

### PR DESCRIPTION
**Pull request by Bhashinee**
## Purpose
Purpose is to add a sample for an online holiday package booking system to demonstrate service orchestration with parallel processing.

## Goals
Demonstrate how to do parallel execution in EI using workers. And to give an idea about how the tasks can be delegated between multiple threads to execute them parallely. 

## Approach
Implemented with a scenario of an online holiday package booking system. Client provides departure date, return date, from, to and the vehicle type he needs in the request. Service will send requests parallely to multiple Airline companies, vehicle rental companies and hotels. Analyzing the responses the most suitable package details will be sent to the client.

### CODE REVIEW

File: samples/holiday-package-sample/services/samples/holidayPackage/HolidayPackageDetails.bal [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+                              };
+        response.setJsonPayload(vehicleDetails);
+        _ = response.send();
+    }
+
+    @http:resourceConfig {
+        methods:["POST"],
+        path:"/airline/sixt"
+    }
+    resource SixT (http:Request request, http:Response response) {
```
Comment (bsenduran) [inline samples/holiday-package-sample/services/samples/holidayPackage/HolidayPackageDetails.bal]: follow a convention in naming the resources, (others are starting in lower cases)

File: samples/holiday-package-sample/services/samples/holidayPackage/HolidayPackage.bal [hunk 2]
```diff
@@ -0,0 +1,265 @@
+package samples.holidayPackage;
```
Comment (bsenduran) [inline samples/holiday-package-sample/services/samples/holidayPackage/HolidayPackage.bal]: Avoid camelCase in package name

File: samples/holiday-package-sample/README.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+}
+```
+
+### Modifying the code
+
+All the services related to this sample is located at <EI_HOME>/samples/holiday-package-sample/services directory.
+
+Following are the available ballerina program files.
+
+[HolidayPackageDetails.bal](services/samples/holidayPackage/HolidayPackageDetails.bal)
```
Comment (bsenduran) [inline samples/holiday-package-sample/README.md]: Better to introduce bulleted list or numbered list

File: samples/holiday-package-sample/services/samples/holidayPackage/HolidayPackage.bal [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+            if (hotelResponses["elizabeth"] != null) {
+                var resElizabethWorker, err = (any[])hotelResponses["elizabeth"];
+                var responseElizabeth, _ = ((http:Response)(resElizabethWorker[0]));
+                elizabethJsonResponse = responseElizabeth.getJsonPayload();
+                elizabethDistance, _ = (int)responseElizabeth.getJsonPayload().DistanceToLocation;
+            }
+
+            if (miramarDistance < aqueenDistance) {
+                if (miramarDistance < elizabethDistance)
+                {
```
Comment (bsenduran) [inline samples/holiday-package-sample/services/samples/holidayPackage/HolidayPackage.bal]: Use `if (...) {`

File: samples/holiday-package-sample/services/samples/holidayPackage/HolidayPackage.bal [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+                    jsonHotelResponse = miramarJsonResponse;
+                }
+            } else {
+                if (aqueenDistance < elizabethDistance) {
+
```
Comment (bsenduran) [inline samples/holiday-package-sample/services/samples/holidayPackage/HolidayPackage.bal]: Unnecessary empty line

File: samples/holiday-package-sample/services/samples/holidayPackage/HolidayPackage.bal [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+        join (some 1) (map airlineResponses) {
+
+            if (airlineResponses["qatarWorker"] != null) {
+                var resQatarWorker, err = (any[])airlineResponses["qatarWorker"];
+                var responseQatar, _ = (http:Response)(resQatarWorker[0]);
+                jsonFlightResponse = responseQatar.getJsonPayload();
+            }
+
+            if (airlineResponses["asianaWorker"] != null) {
+                var resAsianaWorker, err = (any[])airlineResponses["asianaWorker"];
```
Comment (riyafa) [inline samples/holiday-package-sample/services/samples/holidayPackage/HolidayPackage.bal]: This `err` variable if it's not used can't we ignore it. Or we can check this `err` variable before the next line.

File: samples/holiday-package-sample/README.md [hunk 7]
```diff
@@ -0,0 +1,112 @@
+# Online Holiday Packages Booking System.
+
+This scenario is about an online Holiday Package Booking System.
+This is a sample which offers a quick reservation system which involves minimum efforts. This sample will show how
+WSO2 EI is capable of considering a mixture of numerous holiday properties (Flight/ Car/ Hotel) to provide the client with the most suitable holiday package.
```
Comment (isudana) [inline samples/holiday-package-sample/README.md]: I think this description is not correct, we need to explain the scenario first without mentioning about EI, and then describe how we can implement the scenario using EI.

File: samples/holiday-package-sample/services/samples/holiday/HolidayPackage.bal [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+import ballerina.net.http;
+
+@http:configuration {basePath:"/web", port:9092}
+service<http> holidayPackage {
+
+    @http:resourceConfig {
+        methods:["GET"],
+        path:"/holiday"
+    }
+    resource FirstResponding (http:Request request, http:Response response) {
```
Comment (isudana) [inline samples/holiday-package-sample/services/samples/holiday/HolidayPackage.bal]: FirstResponding is not a good resource name

File: samples/holiday-package-sample/services/samples/holiday/HolidayPackage.bal [hunk 9]
```diff
@@ -0,0 +1,263 @@
+package samples.holiday;
+
+import ballerina.net.http;
+
+@http:configuration {basePath:"/web", port:9092}
```
Comment (isudana) [inline samples/holiday-package-sample/services/samples/holiday/HolidayPackage.bal]: Why are we changing the port here ?

File: samples/holiday-package-sample/services/samples/holiday/HolidayPackage.bal [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+                var payload, _ = <json>flightPayload;
+                http:Request req = {};
+                http:Response respWorkerEmirates = {};
+                req.setJsonPayload(payload);
+                respWorkerEmirates, _ = httpEndpoint.post("/detail/airline/emirates", req);
+                respWorkerEmirates -> fork;
+            }
+        }
+
+        join (some 1) (map airlineResponses) {
```
Comment (isudana) [inline samples/holiday-package-sample/services/samples/holiday/HolidayPackage.bal]: please put comment to highlight the important logics.

File: samples/holiday-package-sample/services/samples/holiday/HolidayPackage.bal [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+
+            if (airlineResponses["qatarWorker"] != null) {
+                var resQatarWorker, _ = (any[])airlineResponses["qatarWorker"];
+                var responseQatar, _ = (http:Response)(resQatarWorker[0]);
+                jsonFlightResponse = responseQatar.getJsonPayload();
+            }
+
+            if (airlineResponses["asianaWorker"] != null) {
```
Comment (isudana) [inline samples/holiday-package-sample/services/samples/holiday/HolidayPackage.bal]: shouldn't this be else if ?

File: samples/holiday-package-sample/README.md [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+Client provides departure date, return date, from, to and the vehicle type he needs in the request. Service will send booking requests parallely to multiple Airline companies, vehicle rental companies and hotels.
+Service will include the flight from the first airline to respond and the vehicle from the company which is offering cheapest price to the holiday package.
```
Comment (isudana) [inline samples/holiday-package-sample/README.md]: Shall modify this to get the cheapest airline and first car rental service to respond ?

File: samples/holiday-package-sample/services/samples/holiday/HolidayPackageDetails.bal [hunk 13]
```diff
@@ -0,0 +1,189 @@
+package samples.holiday;
```
Comment (isudana) [inline samples/holiday-package-sample/services/samples/holiday/HolidayPackageDetails.bal]: It doesn't make sense to have a single bal file/service for all the backend services, Please have different services for different entities. We can't even have a single service for same category. All must of individual services. Please think about the real word scenario and mock those services. 
If not it is hard for someone to understand this.

File: samples/holiday-package-sample/services/samples/holiday/HolidayPackage.bal [hunk 14]
```diff
@@ -0,0 +1,263 @@
+package samples.holiday;
```
Comment (erandacr) [inline samples/holiday-package-sample/services/samples/holiday/HolidayPackage.bal]: I think its better if we can add some comments to the Bal files

**Pull request merged**